src/main/java/com/google/sps/data/TextAnalyser.java

```diff
@@ -0,0 +1,169 @@
+    }
+ 
+    return moods[position];
+  }
+
+  private ClassifyTextResponse classify() {
+    try (LanguageServiceClient language = LanguageServiceClient.create()) {
+      Document document = 
+            Document.newBuilder().setContent(message).setType(Type.PLAIN_TEXT).build();
+      ClassifyTextRequest request = ClassifyTextRequest.newBuilder().setDocument(document).build();
+        
+      return language.classifyText(request);
+    } catch(Exception e) {
+      return null;
+    }
+  }
+
+  // list of all the categories that text is about
+  // check https://cloud.google.com/natural-language/docs/categories
+  public List<String> getCategories() {
+    ClassifyTextResponse response = classify();
+
+    if(response == null) {
+      return Collections.emptyList();
+    }
+      
+    ArrayList<String> categories = new ArrayList<String>();
+
+    for (ClassificationCategory category : response.getCategoriesList()) {
+      String[] listCategories = category.getName().split("/");
+      for (int i = 0; i < listCategories.length; i++) {
+        categories.add(listCategories[i].toLowerCase());
+      }
+    }
+
+    return categories;
+  }
+
+  public Set<String> getEvents() {
+    Set<String> events = new HashSet<String>();
+    String[] allEvents = new String[]{"birthday", "wedding", "baby shower", "love",
+                                    "congratulation", "travel", "good morning",
+                                    "gratitude", "job", "promotion",
+                                    "new", "welcome", "good evening", "good night",
+                                    "holiday"};
+    String copy = new String(message);
+    copy = copy.toLowerCase();
+
+    for (int i = 0; i < allEvents.length; i++) {
+      if(copy.indexOf(allEvents[i]) != -1) {
+        events.add(allEvents[i]);
+      }
+    }
+
+    return events;
+  }
+
+  /** Identifies entities in the string */
+  private AnalyzeEntitiesResponse analyzeEntitiesText() throws IOException {
+    // Instantiate the Language client com.google.cloud.language.v1.LanguageServiceClient
+    try (LanguageServiceClient language = LanguageServiceClient.create()) {
+      Document doc = Document.newBuilder().setContent(message).setType(Type.PLAIN_TEXT).build();
+      AnalyzeEntitiesRequest request =
+          AnalyzeEntitiesRequest.newBuilder()
+              .setDocument(doc)
+              .setEncodingType(EncodingType.UTF16)
+              .build();
+
+      return language.analyzeEntities(request);
+    }
+  }
+
+  public List<String> getEntities() throws IOException {
+    AnalyzeEntitiesResponse response = analyzeEntitiesText();
+    List<String> entities = new ArrayList<String>();
+
+    for (Entity entity : response.getEntitiesList()) {
+      entities.add(entity.getName());
+    }
+
+    return entities;
+  }
+
+  // put all the key words together
+  // use a HashSet to remove duplicates
+  public Set<String> getKeyWords() throws IOException {
```

Comment:
Probably this function will be easier to use if it doesn't throw an exception. Generally, I don't know where is the best place (in terms of coding) to catch an exception. Maybe @AndreMB0 and @jeadorf can help us here.

Reply:
I think it would be easier for you if it would catch the exception, so yes, I see your point.

Reply:
What shall the user see if something doesn't work as expected? That I would say is the ultimately judge for the solution. 

More concretely, in this case, what would the alternative be to throwing exceptions? Returning an empty set? How would callers know whether this is a normal expected response (just happens to be empty) or is empty because something went wrong?

More concretely, it seems not possible to resolve and come up with a good action at this level of the code so throwing an exception for some part higher up in the stack to decide what to do seems more appropriate here. I am asking myself here the question: when I catch an exception here, is it going to be actionable─I would argue it isn't here.